src/Microsoft.Identity.Client/TokenCache.Notifications.cs

```diff
@@ -0,0 +1,156 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client
+{
+    public sealed partial class TokenCache : ITokenCacheInternal
+    {
+        /// <summary>
+        /// Notification method called before any library method accesses the cache.
+        /// </summary>
+        internal TokenCacheCallback BeforeAccess { get; set; }
+
+        /// <summary>
+        /// Notification method called before any library method writes to the cache. This notification can be used to reload
+        /// the cache state from a row in database and lock that row. That database row can then be unlocked in the
+        /// <see cref="AfterAccess"/>notification.
+        /// </summary>
+        internal TokenCacheCallback BeforeWrite { get; set; }
+
+        /// <summary>
+        /// Notification method called after any library method accesses the cache.
+        /// </summary>
+        internal TokenCacheCallback AfterAccess { get; set; }
+
+        internal Func<TokenCacheNotificationArgs, Task> AsyncBeforeAccess { get; set; }
+        internal Func<TokenCacheNotificationArgs, Task> AsyncAfterAccess { get; set; }
+        internal Func<TokenCacheNotificationArgs, Task> AsyncBeforeWrite { get; set; }
+
+        internal async Task OnAfterAccessAsync(TokenCacheNotificationArgs args)
+        {
+            AfterAccess?.Invoke(args);
+
+#if !ANDROID_BUILDTIME && !iOS_BUILDTIME
```

Comment:
Not sure why these flags are here. BUILDTIME flags are only to hide public API / signatures. Android runtime will still use the AsyncAfterAccess etc., but the user has no way of setting those on Android.

Reply:
fixed.  i hope we can move to feature flags sometime soon.  our existing #if blocks still stumble me up as to when we use which ones.